### PR TITLE
Change star nomenclature to point source

### LINKED ISSUE #124: Eigen PSF model
**Is your feature request related to a problem? Please describe.**
AstroPhot can fit PSF models live, but we need to be able to account for complex features of emperical PSFs. A good compromise is an Eigen PSF model which could build a PSF as a linear combination of some basis PSF images.

**Describe the solution you'd like**
A PSF model which is just a sum of a set of basis PSF images determined by some other code.

**Describe alternatives you've considered**
It's already possible to fit a parametric PSF or a zernike PSF, but these are more restrictive than the Eigen PSF which could be very general (ie diffraction spikes).

### LINKED ISSUE #125: What's the "approved" way to specify a star (delta function) that should be convolved?
What's the approved way to specify a star?

Currently, there are PSF and star models.  These are meant to be used with PSF convolution _turned off_.  

However, conceptually, a point source is a delta function that one then wants to convolve the same way as other sources in the image.  Performance wise, the convolution and integration (modulo the normalization) may be faster to have already realized in the model; and if that's the answer then I understand.  But it seems a bit prone to error to re-specify the PSF when defining the model.  It seems more symmetric to define a galaxy model and a point source model to be fit together with a known target.PSF.

I'd like to fit a galaxy and a time-variable SN across a set of images.  The SN is a point source; we will assume we have the PSF for the individual images already.

An example model, or an addition to the Model_Zoo tutorial explaining how to simulate a point source.  Or explaining best performance practice is to realize the PSF for the model to fit faster (if that is true).

### PR DESCRIPTION
**Pull request by ConnorStoneAstro**
Closes #125 in principle, though more explanation is likely needed. Using the point source nomenclature should make it more clear that these models are intended for any point source, not just stars.

### CODE REVIEW

**Comment by ConnorStoneAstro** [comment]
perhaps `psf` would be better than `point` as the model type indicator

File: astrophot/models/eigen_psf.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+from ..utils.interpolate import interp2d
+from ._shared_methods import select_target
+from ..param import Param_Unlock, Param_SoftLimits
+from .. import AP_config
+
+__all__ = ["Eigen_PSF"]
+
+class Eigen_PSF(PSF_Model):
+    """point source model which uses multiple images as a basis for the
+    PSF as it's representation for point sources. Using bilinear
```
Comment (wmwv) [inline astrophot/models/eigen_psf.py]:
```suggestion
    PSF as its representation for point sources. Using bilinear
```

File: astrophot/models/eigen_psf.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    interpolation it will shift the PSF within a pixel to accurately
+    represent the center location of a point source. There is no
+    funcitonal form for this object type as any image can be
```
Comment (wmwv) [inline astrophot/models/eigen_psf.py]:
```suggestion
    functional form for this object type as any image can be
```

File: astrophot/models/eigen_psf.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        "flux": {"units": "log10(flux/arcsec^2)"},
+        "weights": {"units": "unitless"},
+    }
+    _parameter_order = PSF_Model._parameter_order + ("flux","weights")
+    useable = True
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if "eigen_basis" not in kwargs:
+            AP_config.ap_logger.warning("Eigen basis not supplied! Assuming psf as single basis element. Please provide Eigen basis or use Pixelated_PSF model.")
```
Comment (wmwv) [inline astrophot/models/eigen_psf.py]: "is a single basis element"?  "has a single basis element"?  "uses a single a basis element"?

File: astrophot/models/pixelated_psf_model.py [hunk 4]
```diff
@@ -1,43 +1,43 @@
...
 
-__all__ = ["PSF_Star"]
+__all__ = ["Pixelated_PSF"]
 
 
-class PSF_Star(Star_Model):
-    """Star model which uses an image of the PSF as it's representation
-    for stars. Using bilinear interpolation it will shift the PSF
+class Pixelated_PSF(PSF_Model):
+    """point source model which uses an image of the PSF as it's representation
```
Comment (wmwv) [inline astrophot/models/pixelated_psf_model.py]:
```suggestion
    """point source model which uses an image of the PSF as its representation
```

File: astrophot/models/psf_model_object.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+from ..utils.decorators import default_internal
+
+
+__all__ = ["PSF_Model"]
+
+
+class PSF_Model(Component_Model):
+    """Prototype point source (typically a star) model, to be subclassed
+    by other point source models which define specific
+    behavior. Mostly this just requires that no standard PSF
```
Comment (wmwv) [inline astrophot/models/psf_model_object.py]: "Mostly this just requires that no standard PSF convolution is applied to this model as that is to be handled internally by the model."

This sentence is a little awkward, and may leave a prospective user confused about who/what is the active agent that would apply a convolution to the model.

File: astrophot/models/sersic_model.py [hunk 6]
```diff
@@ -123,8 +123,8 @@ def _integrate_reference(self, image_data, image_header, parameters):
     from ._shared_methods import sersic_radial_model as radial_model
 
 
-class Sersic_Star(Star_Model):
-    """basic star model with a sersic profile for the radial light
+class Sersic_PSF(PSF_Model):
```
Comment (wmwv) [inline astrophot/models/sersic_model.py:114]: I similarly don't understand what a `Sersic_PSF` is.

**Review by wmwv — COMMENTED** [review]
I think there's some more work to do here in separating the PSF as (A) the thing that is the convolution kernel from the true image to the realized image on the CCD vs. (B) astrophysical things that appear like point sources in images but aren't quite: Nuker, Sersic profiles.

**Comment by ConnorStoneAstro** [comment]
> I think there's some more work to do here in separating the PSF as (A) the thing that is the convolution kernel from the true image to the realized image on the CCD vs. (B) astrophysical things that appear like point sources in images but aren't quite: Nuker, Sersic profiles.

In AstroPhot there isn't really a strong distinction between convolution kernels and point sources in images. For example here: https://autostronomy.github.io/AstroPhot/AdvancedPSFModels.html#PSF-fitting-with-a-bad-star in cell `In [9]` you can see the sersic galaxy model is given a different AstroPhot model (Moffat) to use as it's convolution kernel. You can go further and even use a full group model with multiple components as the convolution kernel. 

There are some unorthodox PSF models in there like the Sersic and Nuker PSF models, but that is really just to give the user freedom in how they want to construct the PSF. Sersic is just a generalization of a Gaussian which is a typical basic PSF model, and Nuker is effectively a double power law. 

So I'm not sure drawing a distinction between the models is the right path, but perhaps I need more explanation on this way of thinking about PSF models?

File: astrophot/models/psf_model_object.py (hunk 5 above)
Comment (ConnorStoneAstro) [reply astrophot/models/psf_model_object.py]: I see what you mean, this doc seems more for myself than a prospective user... I'll update it to make more sense for people other than myself

**Comment by wmwv** [comment]
> So I'm not sure drawing a distinction between the models is the right path, but perhaps I need more explanation on this way of thinking about PSF models?

Hmmm... I see what you're saying mathematically and computationally.

But, conceptually, at least as an observer, I divide these differently in my brain.  The PSF is what convolves that scene that hits top of the atmosphere.  It convolves everything in the image.  The PSF is the same for all objects in an image (with some perhaps smooth spatial variation).

But each object is a different astrophysical object with its own inherent shape.

I think for a new user, in particular, it might be helpful to the user to be pretty clear about the difference between astrophysical model and point source.  I fear that blurring the distinction will make it hard for the user to reason about whether or not they're double-convolving.

One particularly use case: I would like to fit a (let's fixed for now) PSF for the image for all objects.  I thus want to pass a sersic for the galaxy and delta functions for the stars.  These will then get convolved by the PSF.  It's hard for me to reason about how to do that in the current way the models and PSFs are done.

In practice, as AstroPhot is used on more data that has been processed through standard astronomical image pipelines, I think it will want to support Gaussians, Moffats, the families of models from PSFEx, and then also the detailed PSFs from spaced-based missions such as: HST, JWST, GALEX, Swift.

**Comment by ConnorStoneAstro** [comment]
> it might be helpful to the user to be pretty clear about the difference between astrophysical model and point source. I fear that blurring the distinction will make it hard for the user to reason about whether or not they're double-convolving.

Ah, I think I see what you are saying. In principle all the subclasses of `PSF_Model` should actually just be one class called something like `Delta_Function` (probably actually just `Point_Source`) which one then convolves with a PSF to get the image of a point source on the sky. In that sense, all the PSF models (moffat, gaussian, pixelated, eventually PSFEx/HST/JWST/...) should be a whole separate category and not `Component_Model` subclasses at all. Am I understanding your point correctly? In that case I would restructure the models to have the base be `AstroPhot_Model` which is then subclassed by `Component_Model`, `Group_Model`, and `PSF_Model`, all other models would subclass those three appropriately. Does that seem more intuitive?

Right now AstroPhot avoids double convolving internally by forcing convolution to be off for any `PSF_Model` object, but the user may not know this and could justifiably be concerned about double convolution.

**Comment by wmwv** [comment]
> In principle all the subclasses of PSF_Model should actually just be one class called something like Delta_Function (probably actually just Point_Source) which one then convolves with a PSF to get the image of a point source on the sky. In that sense, all the PSF models (moffat, gaussian, pixelated, eventually PSFEx/HST/JWST/...) should be a whole separate category and not Component_Model subclasses at all. Am I understanding your point correctly?

Yes, exactly.

Thank you for being open to re-thinking about the design.

**Comment by ConnorStoneAstro** [comment]
Perfect, I can work on that but it will be a bit slow since I have a number of other things piled up this week. 

I'm always happy to re-structure the design if it makes things clearer for users (assuming its feasible computationally). I really want this project to be widely used :)

**Comment by ConnorStoneAstro** [comment]
So, update on the PSF restructure.

I am building out the new format, but haven't gotten to testing yet. A few things have become clearer along the way.

`PSF_Model` objects will have a `PSF_Image` as their target instead of a `Target_Image` object. This has a few benefits, for one, if you have a stacked empirical PSF then this can be used to initialize the `PSF_Model` just like how a `Component_Model` can use a `Target_Image` to initialize itself. It also handles position information since `PSF_Image` objects are always centered at (0,0) the `PSF_Model` objects don't need to define a center, they can assume (0,0) as the center always. 

The `Point_Source` model which is the delta function subclass of the `Component_Model` now makes more sense in terms of it's meaning than the various star models from before. However, it seems like operationally it is a bit more clunky. To define a point source with a Moffat profile one must now first define a `Moffat_PSF` model which describes the point spread function, then define a `Point_Source` model which actually places a delta function somewhere in the image.  However, I have made it so that when creating a point source you initialize it with a `psf` variable which is a `PSF_Model` object. It would look like:
```
psf = ap.models.AstroPhot_Model(
    model_type = "moffat psf model",
    target = psf_image,
)
star = ap.models.AstroPhot_Model(
    model_type = "pointsource model",
    target = target_image,
    psf = psf,
)
```
whereas before only a single model would be defined which is a `moffat star model`. I have mixed feelings about this since conceptually this is much more clear about what is happening in the code, but operationally it is quite a bit more lines (though the `psf` would only need to be defined once, so it's less noticeable when fitting many point sources). I'd be interested in your thoughts on this.

One thing that I liked about the original system was that in principle any model in AstroPhot could be used to try an model the PSF. In principle this could mean very complex representations, but perhaps that was more confusing than actually helpful. The new system restricts to only use `PSF_model` objects which could potentially be limiting (some bad, some good there). I plan to make a variant of `Group_Model` so that PSF models can be stacked together to make complex representations, but I may end up duplicating a lot of code in order to get the same kind of functionality for both `Component_Model`s and `PSF_Model`s. 

Another complication with the `Point_Source` model is that the core sampling method currently assumes some things that aren't always true anymore. If using a `Pixelated_PSF` model, the pixels do not really represent a smooth analytic model, in an empirical (pixelated) psf the pixels are already integrated by nature of being directly taken from the image. I think this is pretty solvable just by overloading the `sample` method. But this among the reasons listed above means this update won't be super quick to finish... Even for non-pixelated models like the moffat, the way they get sampled is pretty different now. Before, to represent a star with a moffat, I would use the position of the star in the image and integrate directly to the pixels. Now, the moffat model is a psf model and so I sample it into a centered `PSF_Image` which I "convolve" with the delta function by multiplying by the flux, then I do a sub-pixel shift to place the object exactly where it should be. Basically this is the procedure used whenever convolving a PSF with a model, so I already have code for these steps, but it is more computationally expensive than integrating directly to the pixels. I might try and find a work around here just to save compute, not sure yet. 

An idea I had for a completely different approach to this was that instead of making a `PSF_Model` subclass of `AstroPhot_Model` I could have made it so that all models had a "psf mode" which would be activated any time they were provided a `PSF_Image` as their target. This would probably be easier to code since it is quite similar to what I already have implemented. And it would allow using any model as a PSF model. However, it breaks the conceptual clarity and would probably be just as, if not more, confusing than the current system. I also couldn't think of a good name for all the "star" models as they are right now since they would kind of pull double duty as face-on galaxy models, and PSF models. So ultimately I think that would just be more confusing.

Ok, I should probably stop here. Let me know what you think so far! I've pushed the state of the code as it is, nothing works but it might help you see what I'm talking about.

**Review by ConnorStoneAstro — COMMENTED** [review]
need `get_state` / `set_state` for all AstroPhot savable objects. Need `get_fits_state` / `set_fits_state` for all image objects.

**Comment by ConnorStoneAstro** [comment]
Ok, after puttering away for quite a while, I have the PSF update online! PSF models are now a special class of models that take a `PSF_Image` as their target instead of a `Target_Image`. PSF models can be initialized on empirical PSFs just like how regular models can be initialized on target images. A new `Point_Source` model represents a delta function that gets convolved by a PSF to project into an image. I've also cleaned up a bunch of stuff for how PSF images work and how everything interacts. There is a corresponding branch for the tutorials which shows the updates. Let me know what you think!

**Comment by ConnorStoneAstro** [comment]
This also closes #124 as the eigen PSF model was added here as well

**Review by wmwv — APPROVED** [review]
I approve.  I appreciate all the work you put in to think about how this can work from each of the user, conceptual, and technical perspectives.

I offer encouragement to increase the test coverage.

**Comment by wmwv** [comment]
@ConnorStoneAstro Is there more work you're planning to do on the branch for this PR before merging?

**Comment by ConnorStoneAstro** [comment]
> @ConnorStoneAstro Is there more work you're planning to do on the branch for this PR before merging?

@wmwv  Sorry, no life has just gotten away from me. Lets just pull the trigger now and then everyone can enjoy the new updates :)

**Pull request merged**